feat: add callback functionality for binary search tree

dquixote-jr opened the pull request:
I need the functionality to find the k'th node in the tree, which isn't possible with current setup
This augmentation allows people to define whatever functions that they want, in this particular case, it allows me to store the subtree size in a parameter `st_size` and which I can use to find the k'th node in order.

First javascript project PR so this is gonna have some issues, happy to solve them.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/std?pullRequest=6730) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/denoland/std?pullRequest=6730) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/std?pullRequest=6730) it.</sub>

kt3k commented on data_structures/binary_search_tree.ts:
```diff
@@ -104,12 +105,18 @@ export class BinarySearchTree<T> implements Iterable<T> {
    * By default, the values are sorted in ascending order.
    */
-  constructor(compare: (a: T, b: T) => number = ascend) {
+  constructor(compare: (a: T, b: T) => number = ascend, callback?: (node: BinarySearchNode<T>) => void) {
```
Need to document `callback` param in jsdoc.

dquixote-jr replied:
Added the param.

kt3k commented:
> I need the functionality to find the k'th node in the tree, which isn't possible with current setup

This change doesn't seem a good idea for adding that capability to this class. Currently `BinarySearchNode` class is private. This makes it a part of public API. That doesn't seem an ideal change. Can we achieve this without exposing node class public?

For small k's, it seems we can use `lnrValues` (or `rnlValues`) iterator for finding k'th node. What do your tree size and `k` value typically look like?

dquixote-jr commented on data_structures/_test_utils.ts:
```diff
@@ -8,4 +8,5 @@ export class MyMath {
 export interface Container {
   id: number;
   values: number[];
+  st_size: number;
```
done.

dquixote-jr commented:
> For small k's, it seems we can use lnrValues (or rnlValues) iterator for finding k'th node. What do your tree size and k value typically look like?

The tree can be large, like ~5K. And k can be as large as O(n), so iterating over values with lnr/rnl isn't really a good idea.

> This change doesn't seem a good idea for adding that capability to this class. Currently BinarySearchNode class is private. 

I don't think it's the worst thing to make the BinarySearchNode a public class, as we have already made it a "binary" tree, so having left/right/parent elements is expected of it. A different thing to do would be to expose the BinarySearchNode class, but not give access directly to member objects but rather have setChild/getChild or set/get Left/Right/Parent kind of functions.   
I'm open to any other ideas too, my main use case is to allow people to travel over the tree via nodes somehow via left/right nodes. I don't need them to be able to access the specific internal attributes of the class, just left/right children.

kt3k commented:
> I don't think it's the worst thing to make the BinarySearchNode a public class, as we have already made it a "binary" tree, so having left/right/parent elements is expected of it. A different thing to do would be to expose the BinarySearchNode class, but not give access directly to member objects but rather have setChild/getChild or set/get Left/Right/Parent kind of functions.

Then maybe it's better to define a new interface which only has `left`, `right`, and `parent` (and maybe `value`) fields, and use it as input type for callback?

Also run `deno task lint` and see the output. We need to fix them to merge this to main

dquixote-jr commented:
> Then maybe it's better to define a new interface which only has `left`, `right`, and `parent` (and maybe `value`) fields, and use it as input type for callback?

That sounds like a good idea. I've added a new class BSTNode and refactored the setup to use that.

> Also run `deno task lint` and see the output. We need to fix them to merge this to main

Thank you for this. I've got it passing locally. `import type` gave me a headspin but I got it working in the end :)

kt3k commented on data_structures/bst_node.ts:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @typeparam T The type of the values stored in the binary tree.
+ */
+export class BSTNode<T> {
```
We usually avoid using this type of acronym in API names. BinarySearchTree used to be BSTree but we renamed it later https://github.com/denoland/std/pull/2400 I think this should be named `BinarySearchTreeNode`.

Also I think this should be typescript `interface` instead of `class`. We can avoid unnecessary runtime overhead by making this `interface`

dquixote-jr replied:
OK, renamed to BinarySearchTreeNode and made into an interface.

kt3k commented on data_structures/bst_node.ts:
```diff
@@ -0,0 +1,113 @@
+// This module is browser compatible.
+
+/**
+ * A generic Binary Search Tree (BST) node class.
```
Nicely documented 👍

kt3k reviewed (commented):
This now looks much better to me. Left some comments
